src/workerd/api/streams/internal.c++

```diff
@@ -78,35 +78,117 @@ private:
+
     kj::Vector<kj::Array<T>> parts;
     uint64_t runningTotal = 0;
-    static constexpr size_t DEFAULT_BUFFER_CHUNK = 4096;
-    static constexpr size_t MAX_BUFFER_CHUNK = DEFAULT_BUFFER_CHUNK * 4;
+    static constexpr uint64_t DEFAULT_BUFFER_CHUNK = 4096;
+    static constexpr uint64_t MAX_BUFFER_CHUNK = DEFAULT_BUFFER_CHUNK * 4;
 
     // If we know in advance how much data we'll be reading, then we can attempt to
     // optimize the loop here by setting the value specifically so we are only
-    // allocating once. But, to be safe, let's enforce an upper bound on each allocation
-    // even if we do know the total.
-    size_t amountToRead = kj::min(MAX_BUFFER_CHUNK,
-        input.tryGetLength(StreamEncoding::IDENTITY).orDefault(DEFAULT_BUFFER_CHUNK));
-
+    // allocating at most twice. But, to be safe, let's enforce an upper bound on each
+    // allocation even if we do know the total.
+    kj::Maybe<uint64_t> maybeLength = input.tryGetLength(StreamEncoding::IDENTITY);
+
+    // The amountToRead is the regular allocation size we'll use right up until we've
+    // read the number of expected bytes (if known). This number is calculated as the
+    // minimum of (limit, MAX_BUFFER_CHUNK, maybeLength or DEFAULT_BUFFER_CHUNK). In
+    // the best case scenario, this number is calculated such that we can read the
+    // entire stream in one go if the amount of data is small enough and the stream
+    // is well behaved.
+    // If the stream does report a length, once we've read that number of bytes, we'll
+    // fallback to the conservativeAllocation.
+    uint64_t amountToRead = kj::min(limit,
+                                    kj::min(MAX_BUFFER_CHUNK,
+                                            maybeLength.orDefault(DEFAULT_BUFFER_CHUNK)));
+    uint64_t conservativeAllocation = kj::min(amountToRead, DEFAULT_BUFFER_CHUNK);
+
+    // amountToRead can be zero if the stream reported a zero-length. While the stream could
+    // be lying about it's length, let's skip reading anything in this case.
     if (amountToRead > 0) {
       for (;;) {
-        // TODO(perf): We can likely further optimize this loop by checking to see
-        // how much of the buffer was filled and using the remaining buffer space if
-        // it is not completely filled by the previous iteration. Doing so makes this
-        // loop a bit more complicated tho, so for now let's keep things simple.
         auto bytes = kj::heapArray<T>(amountToRead);
-        size_t amount = co_await input.tryRead(bytes.begin(), 1, bytes.size());
+        // Note that we're passing amountToRead as the *minBytes* here so the tryRead should
+        // attempt to fill the entire buffer. If it doesn't, the implication is that we read
+        // everything.
+        uint64_t amount = co_await input.tryRead(bytes.begin(), amountToRead, amountToRead);
+        KJ_DASSERT(amount <= amountToRead);
```

Comment:
New macro to me, reading the source I'm still not clear on `KJ_DASSERT()`, what is different here from `KJ_ASSERT`?

Reply:
Debug assert?